File: golang/src/aoc2020/util/robot/robot.go
package robot

import (
	"aoc2020/util/coord"
)

const NORTH = 0
const EAST = 1
const SOUTH = 2
const WEST = 3

type Direction int

func (d *Direction) TurnLeft() {
	*d = ((*d) - 1) % 4
}

func (d *Direction) TurnRight() {
	*d = ((*d) + 1) % 4
}

func (d Direction) String() string {
	switch d {
	case NORTH:
		return "N"
	case EAST:
		return "E"
	case SOUTH:
		return "S"
	case WEST:
		return "E"
	}
	return "?"
}

type Robot struct {
	Position coord.Coord
	Facing   Direction
}

func (r *Robot) Forward(amount int) {
	var adjust coord.Coord
	switch r.Facing {
	case NORTH:
		adjust = coord.Coord{X: 0, Y: -amount}
	case EAST:
		adjust = coord.Coord{X: amount, Y: 0}
	case SOUTH:
		adjust = coord.Coord{X: 0, Y: amount}
	case WEST:
		adjust = coord.Coord{X: -amount, Y: 0}
	}

	r.Position = coord.Add(r.Position, adjust)
}

func (r *Robot) Backward(amount int) {
	r.Forward(-amount)
}

func (r *Robot) Move(offset coord.Coord) {
	r.Position = coord.Add(r.Position, offset)
}

func (r *Robot) TurnLeft() {
	r.Facing.TurnLeft()
}

func (r *Robot) TurnRight() {
	r.Facing.TurnRight()
}

File: rust/aoc2020/src/util/robot.rs
use crate::util::coord2::{Coord2, Direction};

pub struct Robot {
    pub position: Coord2,
    pub direction: Direction
}

impl Robot {
    pub fn new(dir: Direction) -> Robot {
        Robot {
            position: Coord2{x:0, y:0},
            direction: dir
        }
    }

    pub fn forward(&mut self, amount: i32) {
        let offset = self.direction.coord() * amount;
        self.position = &self.position + &offset;
    }

    pub fn shift(&mut self, dir: Direction, amount: i32) {
        let offset = dir.coord() * amount;
        self.position = &self.position + &offset;
    }

    pub fn shift_coord(&mut self, offset: &Coord2) {
        self.position = &self.position + &offset;
    }

    pub fn move_to(&mut self, pos: &Coord2) {
        self.position = Coord2{x:pos.x, y:pos.y};
    }

    pub fn rotate(&mut self, turns: i32) {
        let idx = self.direction.idx4();
        let new_idx = (((idx as i32 + turns) % 4) + 4) % 4;

        self.direction = Direction::from_idx4(new_idx as usize);
    }

    pub fn manhattan(&self, from: &Coord2) -> i32 {
        self.position.manhattan(from)
    }

}
